net/golioth/fw.c

```diff
@@ -0,0 +1,292 @@
+ */
+
+#include <net/golioth.h>
+#include <net/golioth/fw.h>
+#include <qcbor/qcbor.h>
+#include <qcbor/qcbor_spiffy_decode.h>
+#include <stdio.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(golioth);
+
+#define GOLIOTH_FW_DOWNLOAD	".u"
+#define GOLIOTH_FW_DESIRED	".u/desired"
+
+#define MANIFEST_KEY_SEQUENCE_NUMBER		1
+#define MANIFEST_KEY_HASH			2
+#define MANIFEST_KEY_COMPONENTS			3
+
+#define COMPONENT_KEY_PACKAGE			1
+#define COMPONENT_KEY_VERSION			2
+#define COMPONENT_KEY_HASH			3
+#define COMPONENT_KEY_SIZE			4
+#define COMPONENT_KEY_URI			5
+
+static int parse_component(QCBORDecodeContext *decode_ctx,
+			   uint8_t *version, size_t *version_len,
+			   uint8_t *uri, size_t *uri_len)
+{
+	QCBORError qerr;
+	UsefulBufC version_bufc;
+	UsefulBufC uri_bufc;
+	int err = 0;
+
+	QCBORDecode_EnterMap(decode_ctx, NULL);
+
+	QCBORDecode_GetTextStringInMapN(decode_ctx, COMPONENT_KEY_VERSION,
+					&version_bufc);
+
+	QCBORDecode_GetTextStringInMapN(decode_ctx, COMPONENT_KEY_URI,
+					&uri_bufc);
+
+	qerr = QCBORDecode_GetError(decode_ctx);
+	if (qerr != QCBOR_SUCCESS) {
+		LOG_ERR("Failed to get version and URI: %d (%s)", qerr, qcbor_err_to_str(qerr));
+		err = -EINVAL;
+		goto exit_map;
+	}
+
+	if (version_bufc.len > *version_len) {
+		return -ENOMEM;
+	}
+
+	memcpy(version, version_bufc.ptr, version_bufc.len);
+	*version_len = version_bufc.len;
+
+	if (uri_bufc.len > *uri_len) {
+		return -ENOMEM;
+	}
+
+	memcpy(uri, uri_bufc.ptr, uri_bufc.len);
+	*uri_len = uri_bufc.len;
+
+exit_map:
+	QCBORDecode_ExitMap(decode_ctx);
+
+	return err;
+}
+
+int golioth_fw_desired_parse(const uint8_t *payload, uint16_t payload_len,
+			     uint8_t *version, size_t *version_len,
+			     uint8_t *uri, size_t *uri_len)
+{
+	int64_t manifest_sequence_number;
+	UsefulBufC payload_bufc = { payload, payload_len };
+	QCBORDecodeContext decode_ctx;
+	QCBORError qerr;
+	int err = 0;
+
+	QCBORDecode_Init(&decode_ctx, payload_bufc, QCBOR_DECODE_MODE_NORMAL);
+
+	QCBORDecode_EnterMap(&decode_ctx, NULL);
+
+	QCBORDecode_GetInt64InMapN(&decode_ctx, MANIFEST_KEY_SEQUENCE_NUMBER,
+				   &manifest_sequence_number);
+
+	qerr = QCBORDecode_GetError(&decode_ctx);
+	if (qerr != QCBOR_SUCCESS) {
+		LOG_ERR("Failed to get manifest bstr: %d (%s)", qerr, qcbor_err_to_str(qerr));
+		err = -EINVAL;
+		goto exit_root_map;
+	}
+
+	LOG_INF("Manifest sequence-number: %d", (int) manifest_sequence_number);
+
+	QCBORDecode_EnterArrayFromMapN(&decode_ctx, MANIFEST_KEY_COMPONENTS);
+
+	err = parse_component(&decode_ctx, version, version_len, uri, uri_len);
+
+	QCBORDecode_ExitArray(&decode_ctx);
+
+	if (err) {
+		goto exit_root_map;
+	}
+
+	qerr = QCBORDecode_GetError(&decode_ctx);
+	if (qerr != QCBOR_SUCCESS) {
+		LOG_ERR("Error at the end: %d (%s)", qerr, qcbor_err_to_str(qerr));
+		return -EINVAL;
+	}
+
+exit_root_map:
+	QCBORDecode_ExitMap(&decode_ctx);
+
+	return err;
+}
+
+int golioth_fw_observe_desired(struct golioth_client *client,
+			       struct coap_reply *reply,
```

Comment:
As Zephyr inherits that from Linux kernel, here is an explanation why **not** to typedef structs: https://www.kernel.org/doc/html/v4.14/process/coding-style.html#typedefs

Reply:
I believe that modern IDE's now make it simple enough to find the context of what is behind the typedef.
I like typedefs because they force the user of the type to consider the implications of using that type somewhere it was not originally meant to be used.

While it doesn't 'enforce' stronger typing it at least encourages while also assisting with self documentation of the code.

Reply:
Keep with convention for now. I'm not going to shake things up just yet.